### ISSUE TITLE
Regression calling POJO setters twice from DefaultRecordMapper.MutablePOJOMapper

### ISSUE THREAD

**Issue by jecker-turning** [jOOQ/jOOQ#8460]
### Expected behavior and actual behavior:
Expected behavior is that the setter for an object is called only once per row.
Actual behavior is that it is called twice per row.

There probably isn't too much code out there that would be flat out broken by this, but it seems a bit off.

I've tested this behavior with 3.11.5, and it did not happen then. It still appears to occur with 3.11.10.

### Steps to reproduce the problem:
Create a simple class like:
```
class SimpleTest
{
    public int totalTimesCalled = 0;

    public void setTest(int i)
    {
        totalTimesCalled++;
    }
}
```

Run a simple query that only returns one record, like this:
```//jooq is a DSLContext )Object
SimpleTest test = jooq.fetchOne("SELECT 1 AS test").into(SimpleTest.class);
```
You will see that `totalTimesCalled` is set to 2 rather than 1.

### Versions:

- jOOQ: 3.11.6.
- Java: 8
- Database (include vendor): postgres 9.5
- OS: macOS 10.14
- JDBC Driver (include name if inofficial driver): org.postgresql.Driver version 42.2.5

**Comment by lukaseder** [comment]
Thank you very much for your report. That's sounds ... intriguing :) But I cannot reproduce this with 3.11.10 or 3.12 using this test case:

```java

    @Test
    public void testMutablePojo() {
        Field<String> a = field("a", VARCHAR);

        Record1<String> rec = create.newRecord(a);
        rec.values("a");
        MutableWithSetter pojo = rec.into(MutableWithSetter.class);

        assertEquals(1, pojo.timesCalled);
        assertEquals("a", pojo.concat);
    }

    static class MutableWithSetter {
        int timesCalled;
        String concat = "";

        void setA(String a) {
            timesCalled++;
            concat += a;
        }
    }
```

Would you mind creating an MCVE to help reproduce this problem? 
https://github.com/jOOQ/jOOQ-mcve

**Comment by jecker-turning** [comment]
Here's the MCVE: https://github.com/jecker-turning/jOOQ-mcve

Thanks for taking the time to look at this Lukas! I love the work you guys do with jooq.

**Comment by lukaseder** [comment]
OK, I can reproduce it with your MCVE. Should have tried directly as you suggested instead of "simplifying" it. Will investigate now

**Comment by lukaseder** [comment]
This seems to be a regression introduced with #7953. This internal reflection logic in the `Tools` class can lead to duplicate methods in the result list:

```java
    private static final List<Method> getInstanceMethods(Class<?> type) {
        List<Method> result = new ArrayList<Method>();

        for (Method method : type.getMethods())
            if ((method.getModifiers() & Modifier.STATIC) == 0)
                result.add(method);

        do {
            for (Method method : type.getDeclaredMethods())
                if ((method.getModifiers() & Modifier.STATIC) == 0)
                    result.add(method);

            type = type.getSuperclass();
        }
        while (type != null);

        return result;
    }
```

I couldn't detect the regression in my test case because my class `MutableWithSetter` class is not public, so `MutableWithSetter.class.getMethods()` doesn't return the `setA()` method.

**Comment by lukaseder** [comment]
This test does reproduce the problem:

```java

    @Test
    public void testMutablePojo() {
        // [#7953] Make sure setters are called only once

        Field<String> a = field("a", VARCHAR);

        Record1<String> rec = create.newRecord(a);
        rec.values("a");

        PrivateMutableWithSetter p1 = rec.into(PrivateMutableWithSetter.class);
        assertEquals(1, p1.timesCalled);
        assertEquals("a", p1.concat);

        PublicMutableWithSetter p2 = rec.into(PublicMutableWithSetter.class);
        assertEquals(1, p2.timesCalled);
        assertEquals("a", p2.concat);
    }

    static class PrivateMutableWithSetter {
        int timesCalled;
        String concat = "";

        void setA(String a) {
            timesCalled++;
            concat += a;
        }
    }

    public static class PublicMutableWithSetter {
        int timesCalled;
        String concat = "";

        public void setA(String a) {
            timesCalled++;
            concat += a;
        }
    }
```

**Comment by lukaseder** [comment]
We could put the methods in a `Set` and return that instead. But with the algorithm that finds declared methods on all super types, this could still result in duplicate methods:

1. If one of the methods is a synthetic method, e.g. a bridge method
2. If one of the methods overrides a method from a super type

We might need a wrapper object that implements `equals()` and `hashCode()` based on the source signature (name and parameter types), not the binary signature (declaring class, return type, name, and parameter types)

**Comment by lukaseder** [comment]
Care needs to be applied here. While we never want duplicate identical methods, we sometimes want duplicate methods by signature, as annotations are not inherited, for example. So if a super class annotates a method and a subclass overrides it without repeating the annotation, we need to check both methods for the annotation's presence.

**Comment by lukaseder** [comment]
This test also needs to work:

```java

    public static class PublicMutableWithSetter {
        int timesCalled;
        String concat = "";

        public void setA(String a) {
            timesCalled++;
            concat += a;
        }
    }

    public static class SubtypeMutableWithSetter extends PublicMutableWithSetter{
        @Override
        public void setA(String a) {
            super.setA(a);
        }
    }
```

And then:

```java
        SubtypeMutableWithSetter p3 = rec.into(SubtypeMutableWithSetter.class);
        assertEquals(1, p3.timesCalled);
        assertEquals("a", p3.concat);
```

**Comment by lukaseder** [comment]
The overhead of copying around methods in sets and calculating additional `equals()` and `hashCode()` values can be ignored since we're caching the results already on a per `(type, field name)` basis.

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.12. Backport scheduled for 3.11.11 (#8472) and 3.10.9 (#8473)
